packages/zoe/src/contracts/attestation/expiringNFT.js

```diff
@@ -0,0 +1,264 @@
+ *
+ * @param {Address} address
+ * @param {Amount} amountLiened - the amount of the underlying asset to put
+ * a lien on
+ * @param {Timestamp} expiration
+ * @param {Handle<'attestation'>} handle - the unique handle
+ * @returns {ExpiringAttElem}
+ */
+const makeAttestationElem = (address, amountLiened, expiration, handle) => {
+  return harden({
+    address,
+    amountLiened,
+    expiration,
+    handle,
+  });
+};
+
+/**
+ * We should be as reluctant as possible to lift a lien, therefore,
+ * the expiration has only expired when the currentTime is past
+ * (greater than and not equal to) the expiration.
+ *
+ * @param {Timestamp} expiration
+ * @param {Timestamp} currentTime
+ * @returns {boolean}
+ */
+const hasExpired = (expiration, currentTime) => expiration < currentTime;
+
+/**
+ * @param {string} attestationTokenName - the name for the attestation
+ * token
+ * @param {Amount} empty - an empty amount in the external brand (i.e.
+ * BLD) that the attestation is about
+ * @param {ContractFacet} zcf
+ * @returns {Promise<{disallowExtensions: DisallowExtensions, addExpiringLien: AddExpiringLien, getLienAmount:
+ * getLienAmount, extendExpiration: ExtendExpiration, issuer: Issuer, brand: Brand}>}
+ */
+const setupAttestation = async (attestationTokenName, empty, zcf) => {
+  const zcfMint = await zcf.makeZCFMint(attestationTokenName, AssetKind.SET);
+  const {
+    brand: attestationBrand,
+    issuer: attestationIssuer,
+  } = zcfMint.getIssuerRecord();
+
+  const externalBrand = empty.brand;
+
+  // Note: `amountLiened` is of the brand `externalBrand`
+
+  /** @type {Store<Address,Array<ExpiringAttElem>>} */
+  const lienedAmounts = makeStore('address');
+
+  const cannotGetExtension = new Set();
+
+  /**
+   * Add a lien for a particular amount for an address, with an
+   * expiration date after which it is unliened (whether the
+   * expiration date has passed is only checked when `getLienAmount`
+   * is called)
+   *
+   * @param {ExpiringAttElem} attestationElem
+   */
+  const addToLiened = attestationElem => {
+    const { address } = attestationElem;
+    if (lienedAmounts.has(address)) {
+      const lienedSoFar = lienedAmounts.get(address);
+      lienedSoFar.push(attestationElem);
+      lienedAmounts.set(address, lienedSoFar);
+    } else {
+      lienedAmounts.init(address, [attestationElem]);
+    }
+  };
+
+  // This request *must* come from the owner of the address. Merely
+  // verifying that the address *could* add a lien is not sufficient.
+  // The owner must consent to adding a lien, and non-owners must not
+  // be able to initiate a lien for another account.
+
+  /** @type {AddExpiringLien} */
+  const addExpiringLien = (address, amount, expiration) => {
+    const amountToLien = validateInputs(externalBrand, address, amount);
+    assert.typeof(expiration, 'bigint');
+
+    const handle = makeHandle('attestation');
+
+    const attestationElem = makeAttestationElem(
+      address,
```

Comment:
The Returnable Attestations require the address so that the attestation contract knows which address can be credited for a returned attestation. I think the attestations should have a common format.